pandas/io/html.py

```diff
@@ -637,41 +742,32 @@ def _build_doc(self):
                 raise XMLSyntaxError("no text parsed from document", 0, 0, 0)
         return r
 
-    def _parse_tbody(self, table):
-        return table.xpath('.//tbody')
-
-    def _parse_thead(self, table):
-        return table.xpath('.//thead')
-
-    def _parse_tfoot(self, table):
-        return table.xpath('.//tfoot')
-
-    def _parse_raw_thead(self, table):
-        expr = './/thead'
-        thead = table.xpath(expr)
-        res = []
-        if thead:
-            # Grab any directly descending table headers first
-            ths = thead[0].xpath('./th')
-            if ths:
-                cols = [_remove_whitespace(x.text_content()) for x in ths]
-                if any(col != '' for col in cols):
-                    res.append(cols)
-            else:
-                trs = self._parse_tr(thead[0])
+    def _parse_thead_tr(self, table):
+        rows = []
 
-                for tr in trs:
-                    cols = [_remove_whitespace(x.text_content()) for x in
-                            self._parse_td(tr)]
+        for thead in table.xpath('.//thead'):
+            rows.extend(thead.xpath('./tr'))
 
-                    if any(col != '' for col in cols):
-                        res.append(cols)
-        return res
+            # HACK: lxml does not clean up the clearly-erroneous
```

Comment:
The comment here mentions a situation we are missing `tr` elements but `test_header_inferred_from_th_elements` is constructed without a `thead` element instead - is that by design?

Reply:
That test is unrelated to this hack. This hack is tested in `test_thead_without_tr`.